### DISCUSSION TITLE
Csla 6 fluent configuration - should there be a RegisterContextManager extension method?

### DISCUSSION THREAD

**Discussion by TheCakeMonster** [MarimerLLC/csla#2699] [Ideas]
Csla contains cleverness to automatically select the right ApplicationContextManager for the environment within which it is running. This almost always just works, and doesn't need configuring. However, there are edge cases, and one of them is unit testing.

It is currently possible to specify a custom context manager by registering your own with DI, but you need to know to register it against the IContextManager interface for it to work.
```csharp
services.AddScoped<IContextManager, MyCustomContextManager>();
```
We could add an extension method, probably to CslaOptions, to make this a bit more discoverable, so that it can be done via the AddCsla method
```csharp
services.AddCsla(options => options.RegisterContextManager<MyCustomContextManager>());
```
I suppose there is a chance that this could encourage people with limited knowledge to change it out and then raise support issues and questions about why something isn't working as intended. However, I think the easier discoverability of it is probably worth this potential cost of answering questions in the unlikely event that someone changes it incorrectly.

**Comment by rockfordlhotka** [comment]
I like the AddContextManager name - doesn't that fit better with what Microsoft does? In any case, this is a good suggestion for sure.

**Reply by TheCakeMonster** [reply]
Yep, we could use Add. I only suggested Register because that's what we used for everything in DataPortalServerOptions for situations in which there is only support for a single one. We used Add and Remove for interceptors because there can be more than one - we're affecting a list.

**Reply by rockfordlhotka** [reply]
Good point, as long as we're _internally_ consistent.

**Reply by swegele** [reply]
@TheCakeMonster - Thinking about your comment:

> there are edge cases, and one of them is unit testing

I wonder if it would be valuable to have a Csla demo project that documents, for other users, examples of what you mention.  In it you could write a few different unit tests (say via MSTest v2 framework)?  Or it might also work to add a testing project to one of the existing demos as an additional project.  
I know each demo brings with it maintenance overhead, but then again, this is arguably something we "should" be doing if we have a business layer so maybe good to give a breadcrumb path.
Personally, I always look to the demo projects for guidance, assuming that it shows, if not the best, at least a good way, to do things.

**Reply by TheCakeMonster** [reply]
Yeah, that might make sense. However, now that you've made me think about it, that might form part of a separate, and slightly bigger, discussion. **Incoming**!
